File: Dished/Layout/DAProgressView.m
//
//  DAProgressView.m
//  Dished
//
//  Created by Ryan Khalili on 9/11/14.
//  Copyright (c) 2014 Dished. All rights reserved.
//

#import "DAProgressView.h"


@interface DAProgressView()

@property (strong, nonatomic) CALayer      *maskLayer;
@property (strong, nonatomic) CALayer      *blueDishLayer;
@property (strong, nonatomic) CALayer      *grayDishLayer;

@end


@implementation DAProgressView

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    
    if( self )
    {
        [self setupImageLayer];
    }
    
    return self;
}

- (void)setupImageLayer
{
    self.grayDishLayer = [CALayer layer];
    self.grayDishLayer.masksToBounds = YES;
    UIImage *dishImage = [UIImage imageNamed:@"refresh_gray"];
    CGFloat x = ( self.frame.size.width  / 2 ) - ( dishImage.size.width  / 2 );
    CGFloat y = ( self.frame.size.height / 2 ) - ( dishImage.size.height / 2 );
    self.grayDishLayer.frame = CGRectMake( x, y, dishImage.size.width, dishImage.size.height );
    self.grayDishLayer.backgroundColor = [UIColor clearColor].CGColor;
    self.grayDishLayer.contents = (id)dishImage.CGImage;
    [self.layer addSublayer:self.grayDishLayer];
    
    self.blueDishLayer = [CALayer layer];
    self.blueDishLayer.frame = self.grayDishLayer.frame;
    self.blueDishLayer.backgroundColor = [UIColor clearColor].CGColor;
    self.blueDishLayer.contents = (id)[UIImage imageNamed:@"refresh_blue"].CGImage;
    [self.layer addSublayer:self.blueDishLayer];
    
    self.maskLayer = [CALayer layer];
    self.maskLayer.anchorPoint = CGPointZero;
    self.maskLayer.frame = CGRectMake( 0, 0, 0, self.blueDishLayer.frame.size.height );
    self.maskLayer.backgroundColor = [UIColor blackColor].CGColor;
    self.blueDishLayer.mask = self.maskLayer;
}

- (void)animateToPercentage:(CGFloat)percentage
{
    [self.maskLayer removeAllAnimations];
    
    if( percentage != percentage )
    {
        return;
    }
    
    CGFloat width = self.blueDishLayer.frame.size.width * percentage;
    CGRect maskFrame = self.maskLayer.frame;
    maskFrame.size.width = width;
    
    if( width != width )
    {
        return;
    }
    
    CABasicAnimation *progressAnimation = [CABasicAnimation animationWithKeyPath:@"bounds.size.width"];
    progressAnimation.fromValue = @(self.maskLayer.frame.size.width);
    progressAnimation.toValue   = @(width);
    progressAnimation.duration  = 0.1;
    self.maskLayer.frame = maskFrame;
    [self.maskLayer addAnimation:progressAnimation forKey:@"progress"];
}

@end

File: Dished/Layout/DAProgressView.swift
//
//  DAProgressView2.swift
//  Dished
//
//  Created by Ryan Khalili on 3/21/15.
//  Copyright (c) 2015 Dished. All rights reserved.
//

import UIKit

class DAProgressView: DAView {
    
    private var maskLayer: CALayer!
    private var grayDishLayer: CALayer!
    private var blueDishLayer: CALayer!

    override func setupViews() {
        setupImageLayer()
    }
    
    private func setupImageLayer() {
        grayDishLayer = CALayer()
        grayDishLayer.masksToBounds = true
        let dishImage = UIImage(named: "refresh_gray")!
        let x = (frame.size.width / 2) - (dishImage.size.width / 2)
        let y = (frame.size.height / 2) - (dishImage.size.height / 2)
        grayDishLayer.frame = CGRectMake(x, y, dishImage.size.width, dishImage.size.height)
        grayDishLayer.backgroundColor = UIColor.clearColor().CGColor
        grayDishLayer.contents = dishImage.CGImage
        layer.addSublayer(grayDishLayer)
        
        blueDishLayer = CALayer()
        blueDishLayer.frame = grayDishLayer.frame
        blueDishLayer.backgroundColor = UIColor.clearColor().CGColor
        blueDishLayer.contents = UIImage(named: "refresh_blue")!.CGImage
        layer.addSublayer(blueDishLayer)
        
        maskLayer = CALayer()
        maskLayer.anchorPoint = CGPointZero
        maskLayer.frame = CGRectMake(0, 0, 0, blueDishLayer.frame.size.height)
        maskLayer.backgroundColor = UIColor.blackColor().CGColor
        blueDishLayer.mask = maskLayer
    }
    
    func animateToPercentage(percentage: Float) {
        maskLayer.removeAllAnimations()
        
        if percentage.isNaN {
            return
        }
        
        let width = Float(blueDishLayer.frame.size.width) * percentage
        var maskFrame = maskLayer.frame
        maskFrame.size.width = CGFloat(width)
        
        if width.isNaN {
            return
        }
        
        let animation = CABasicAnimation(keyPath: "bounds.size.width")
        animation.fromValue = maskLayer.frame.size.width
        animation.toValue = width
        animation.duration = 0.1
        maskLayer.frame = maskFrame
        maskLayer.addAnimation(animation, forKey: "progress")
    }
}